onnx/defs/math/defs.cc

```diff
@@ -3660,4 +3660,534 @@ ONNX_OPERATOR_SET_SCHEMA(
+        OpSchema()
+            .SetDoc(DFT_ver13_doc)
+            .Attr(
+                "onesided",
+                "If onesided is 1, only values for ω in [0, 1, 2, ..., floor(n_fft/2) + 1] are returned because "
+                "the real-to-complex Fourier transform satisfies the conjugate symmetry, i.e., X[m, ω] = X[m,ω]=X[m,n_fft−ω]*. "
+                "Note if the input or window tensors are complex, then onesided output is not possible. "
+                "Enabling onesided with real inputs performs a Real-valued fast Fourier transform (RFFT)."
+                "When invoked with real or complex valued input, the default value is 0. "
+                "Values can be 0 or 1.",
+                AttributeProto::INT,
+                static_cast<int64_t>(0))
+            .Input(
+                0,
+                "input",
+                "Input tensor representing a real or complex valued signal. "
+                "For real input, the following shape is expected: [batch_size][n_fft]. "
+                "For complex input, the following shape is expected: [batch_size][n_fft][2], where " 
+                "[batch_size][n_fft][0] represents the real component and [batch_size][n_fft][1] represents the imaginary component of the signal.",
+                "T",
+                OpSchema::Single,
+                true,
+                1,
+                OpSchema::NonDifferentiable)
+            .Output(
+                0,
+                "output",
+                "The discrete Fourier transform of input. "
+                "If onesided is 1, the output has the shape: [batch_size][floor(n_fft/2)+1][2]. " 
+                "If onesided is 0, the output has the shape: [batch_size][n_fft][2]"
+                "For all types of input, the last dimension of the output represents the components of a complex number.",
+                "T")
+            .TypeConstraint(
+                "T",
+                {"tensor(float16)", "tensor(float)", "tensor(double)"},
+                "Constrain input and output types to float tensors.")
+            .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+                propagateElemTypeFromInputToOutput(ctx, 0, 0);
+                // To support multidimensional DFT the signal_ndim may != 1
+                const int64_t batch_ndim = 1;
+                const int64_t component_ndim = 1;
+
+                // To support multidimensional DFT the signal_ndim may != 1,
+                // but this attribute is not included on the schema yet
+                const auto signal_ndim = 1;
+
+                // Add 1 dimensions to the rank for the batch size dimension
+                auto real_inputs_ndim = signal_ndim + batch_ndim;
+                // Add 2 dimensions to the rank for the batch size dimension AND component dimension
+                auto complex_inputs_ndim = signal_ndim + batch_ndim + component_ndim;
+
+                auto& input_shape = getInputShape(ctx, 0);
```

Comment:
Best to make all the code from here conditional on ```hasInputShape(ctx,0)```, and check that the shape has the expected dimensions.

Reply:
Hi, the shape inference code has issues that must be addressed, since we cannot assume that the rank/dimensions of an input are known during shape-inference time. Please see this PR who updated documentation on shape-inference to help clarify a few things. Please do let me know if you have any questions/suggestions. Thanks!

Reply:
Forgot the PR link: https://github.com/onnx/onnx/pull/4163

Reply:
Updated to include hasInputShape checks.